zebra-chain/src/primitives/zcash_history.rs

```diff
@@ -0,0 +1,329 @@
+    let work = bigint::U256::from_big_endian(&work.as_u128().to_be_bytes());
+
+    let sapling_tx_count = count_sapling_transactions(block);
+
+    zcash_history::NodeData {
+        consensus_branch_id: branch_id.into(),
+        subtree_commitment: block_hash,
+        start_time: time,
+        end_time: time,
+        start_target: target,
+        end_target: target,
+        start_sapling_root: sapling_root,
+        end_sapling_root: sapling_root,
+        subtree_total_work: work,
+        start_height: height.0 as u64,
+        end_height: height.0 as u64,
+        sapling_tx: sapling_tx_count,
+    }
+}
+
+/// Count how many Sapling transactions exist in a block,
+/// i.e. transactions "where either of vSpendsSapling or vOutputsSapling is non-empty"
+/// (https://zips.z.cash/zip-0221#tree-node-specification).
+fn count_sapling_transactions(block: Arc<Block>) -> u64 {
+    block
+        .transactions
+        .iter()
+        .filter(|tx| tx.has_sapling_shielded_data())
+        .count()
+        .try_into()
+        .expect("number of transactions must fit u64")
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        block::Commitment::{self, ChainHistoryActivationReserved},
+        serialization::ZcashDeserializeInto,
+    };
+
+    use super::*;
+    use color_eyre::eyre;
+    use eyre::Result;
+    use zebra_test::vectors::{
+        MAINNET_BLOCKS, MAINNET_FINAL_SAPLING_ROOTS, TESTNET_BLOCKS, TESTNET_FINAL_SAPLING_ROOTS,
+    };
+
+    #[test]
+    fn tree() -> Result<()> {
+        tree_for_network(Network::Mainnet, 903_000)?;
+        tree_for_network(Network::Testnet, 903_800)?;
+        Ok(())
+    }
+
+    fn tree_for_network(network: Network, height: u32) -> Result<()> {
+        let (blocks, sapling_roots) = match network {
+            Network::Mainnet => (&*MAINNET_BLOCKS, &*MAINNET_FINAL_SAPLING_ROOTS),
+            Network::Testnet => (&*TESTNET_BLOCKS, &*TESTNET_FINAL_SAPLING_ROOTS),
+        };
+
+        // Load Block 0 (Heartwood activation block)
+        let block0 = Arc::new(
+            blocks
+                .get(&height)
+                .expect("test vector exists")
+                .zcash_deserialize_into::<Block>()
+                .expect("block is structurally valid"),
+        );
+
+        // Check its commitment
+        let commitment0 = block0.commitment(network)?;
+        assert_eq!(commitment0, ChainHistoryActivationReserved);
+
+        // Build initial MMR tree with only Block 0
+        let sapling_root0 =
+            sapling::tree::Root(**sapling_roots.get(&height).expect("test vector exists"));
+        let entry0 = Entry::new_leaf(block0, network, &sapling_root0);
+        let mut peaks = HashMap::new();
+        peaks.insert(0u32, &entry0);
+        let mut tree = Tree::new(
+            network,
+            NetworkUpgrade::Heartwood,
+            1,
+            &peaks,
+            &HashMap::new(),
+        )?;
```

Comment:
Can we move all this code into `Tree::new`?
It seems like we'll have to do it each time we create a new tree at the start of each network upgrade?
Can we split `new` into `new_from_cache` and `new_empty`?

Reply:
I think the wrapper function looks good.